Drop Crate on Death

FS-21 opened the pull request:
- Ìf `DropCrate` is declared then the specified crate will be spawned when the object is destroyed.

In `rulesmd.ini`:
```ini
[SOMETECHNO]  ; TechnoType
DropCrate=    ; Powerup crate type enum (money|unit|healbase|cloak|explosion|napalm|squad|reveal|armor|speed|firepower|icbm|invulnerability|veteran|ionstorm|gas|tiberium|pod)
```

Also, a new map action (see the documentation) allows the overwrite of this behaviour (disbale, enable or set a new crate) if the trigger is attached to technos.

Tested in online games so this doesn't desync.

FS-21 committed:
Initial commit

- Ìf `DropCrate` is declared then the specified crate will be spawned when the object is destroyed.

In `rulesmd.ini`:
```ini
[SOMETECHNO]  ; TechnoType
DropCrate=    ; Powerup crate type enum (money|unit|healbase|cloak|explosion|napalm|squad|reveal|armor|speed|firepower|icbm|invulnerability|veteran|ionstorm|gas|tiberium|pod)
```

Also, a new map action (see the documentation) allows the overwrite of this behaviour (or set as new value!) if the trigger is attached to technos.

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/dropcrate

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/dropcrate

FS-21 committed:
add documentation image

FS-21 committed:
Merge branch 'develop' into feature/dropcrate

MortonPL reviewed (commented):
Won't collecting this new crate spawn a new one according to random crate logic?

FS-21 committed:
merge from develop

FS-21 committed:
Merge branch 'develop' into feature/dropcrate

FS-21 committed:
Merge branch 'develop' into feature/dropcrate

FS-21 committed:
Merge branch 'develop' into feature/dropcrate

Coronia committed:
Merge remote-tracking branch 'upstream/develop' into feature/dropcrate

# Conflicts:
#	CREDITS.md
#	YRpp
#	docs/New-or-Enhanced-Logics.md
#	src/Ext/Techno/Body.cpp
#	src/Ext/Techno/Body.h
#	src/Ext/TechnoType/Body.cpp
#	src/Ext/TechnoType/Body.h

Coronia committed:
Apply suggestions from code review

Co-authored-by: Bartłomiej Moroz <Bartm12@wp.pl>

Coronia commented on src/Ext/Techno/Hooks.cpp:
```diff
@@ -360,6 +360,31 @@ DEFINE_HOOK(0x4DB218, FootClass_GetMovementSpeed_SpeedMultiplier, 0x6)
+		Powerup selectedPowerup = static_cast<Powerup>(nSelectedPowerup);
+		TechnoExt::TryToCreateCrate(pThis->Location, selectedPowerup);
+	}
+
```
the entire section should be put under a proper techno death handling hook in `Hooks.ReceiveDamage`, otherwise it might have the chance to miss or duplicate the crate

FS-21 replied:
First of all, why all the code calling this feature was moved from the hook "DEFINE_HOOK(0x701DFF, TechnoClass_ReceiveDamage_FlyingStrings, 0x7)" (initial commit I created) to "DEFINE_HOOK(0x4DB218, FootClass_GetMovementSpeed_SpeedMultiplier, 0x6)" (current commit)...

Now I understand the code started to drop random instead of a specific one stuff some time ago.
Originally it was as you mentioned Coronia. I'll push a fix moving the code to the original hook.

FS-21 replied:
Now should be fixed

FS-21 committed:
Move code into the original hook

Maybe some merge moved the code, I have no idea.

FS-21 committed:
Merge branch 'develop' into feature/dropcrate

TaranDahl reviewed (changes requested):
Two concerns.

FS-21 committed:
Merge branch 'develop' into feature/dropcrate

FS-21 committed:
Fixed problem with Temporal kills

- Moved code to another function that can detect Temporal deaths.
- Improved code.

FS-21 commented on src/Ext/Techno/Body.h:
```diff
@@ -71,6 +71,9 @@ class TechnoExt
 
 		AirstrikeClass* AirstrikeTargetingMe;
 
+		int DropCrate; // Drop crate on death, modified by map action
```
Ehh, this PR has map actions that modify the behaviour of drop crates enabling/disabling and modifying the current value

FS-21 committed:
Merge branch 'develop' into feature/dropcrate

FS-21 committed:
Merge branch 'develop' into feature/dropcrate

FS-21 committed:
Merge branch 'develop' into feature/dropcrate

Starkku commented:
Is there any existing vanilla function that is suitable for finding available cells for crate spawning instead of implementing your own code?

FS-21 committed:
Merge branch 'develop' into feature/dropcrate

FS-21 committed:
Merge branch 'develop' into feature/dropcrate

FS-21 committed:
Merge branch 'develop' into feature/dropcrate

Coronia committed:
Merge branch 'develop' into feature/dropcrate

# Conflicts:
#	docs/AI-Scripting-and-Mapping.md
#	docs/Whats-New.md
#	src/Ext/TAction/Body.cpp
#	src/Ext/TAction/Body.h
#	src/Ext/Techno/Body.cpp
#	src/Ext/Techno/Body.h
#	src/Ext/TechnoType/Body.cpp
#	src/Ext/TechnoType/Body.h

Coronia committed:
Merge branch 'develop' into feature/dropcrate

Coronia committed:
fix doc

Coronia committed:
doc

Coronia committed:
const
